pkg/addon/addon.go

```diff
@@ -1072,6 +1092,221 @@ func (h *Installer) installDependency(addon *InstallPackage) error {
 	return nil
 }
 
+// validateAddonDependencies checks if addon's dependencies can be satisfied.
+// If dependency is installed, check if the version matches required version.
+// If dependency is not installed, check available addons for dependency
+// matching the required version.
+// Return error if any dependency cannot be satisfied.
+func validateAddonDependencies(addon *InstallPackage, installedAddons addonInfoMap, availableAddons addonInfoMap) error {
+	dependenciesInstalled := []Dependency{}
+	dependenciesNotInstalled := []Dependency{}
+	for _, dep := range addon.Dependencies {
+		_, ok := installedAddons[dep.Name]
+		if ok {
+			dependenciesInstalled = append(dependenciesInstalled, *dep)
+		} else {
+			dependenciesNotInstalled = append(dependenciesNotInstalled, *dep)
+		}
+	}
+
+	// Check dependencies against installed addon versions.
+	// If dependency version is specified, check if installed version matches
+	// required version.
+	var err error
+	for _, dep := range dependenciesInstalled {
+		if dep.Version == "" {
+			continue
+		}
+
+		// get installed version
+		installedAddon := installedAddons[dep.Name]
+		// versions length must be 1
+		if len(installedAddon.AvailableVersions) != 1 {
+			err = multierr.Append(err, errors.New("installedAddon.Versions length must be 1"))
+			continue
+		}
+		installedVersion := installedAddon.AvailableVersions[0]
+
+		match, _ := checkSemVer(installedVersion, dep.Version)
+		if !match {
+			err = multierr.Append(err, fmt.Errorf(
+				"addon %s has unresolvable dependency %s, required version '%s', installed version '%s'",
+				addon.Name, dep.Name, dep.Version, installedVersion))
+		}
+	}
+
+	// Check dependencies not installed against available addon versions.
+	// If dependency version is specified, check if available version matches
+	// required version.
+	for _, dep := range dependenciesNotInstalled {
+		availableAddon, ok := availableAddons[dep.Name]
+		if !ok {
+			err = multierr.Append(err, fmt.Errorf(
+				"addon %s has unresolvable dependency %s, required version '%s', no available addon with name",
+				addon.Name, dep.Name, dep.Version))
+		} else {
+			if dep.Version == "" {
+				continue
+			}
+
+			_, err2 := calculateDependencyVersionToInstall(dep, nil, availableAddons)
```

Comment:
Maybe we could return a variable that records the version of each dependency addon that needs to be installed, so we don't have to run this method again to get the version that the addon later.
Thus, The logic that installedAddons are not nil  in calculateDependencyVersionToInstall function is also not needed.

Reply:
That's true, we could do this. I worried that this might make the validation function less clean since it would have 2 separate responsibilities:
1. Validate dependencies
2. Calculate dependency versions to install

Reply:
This concern is warranted. If we can use to replace validateAddonDependencies calculateDependencyVersionToInstall directly, ValidateAddonDependencies logic is also have to look for the appropriate version can be installed. we can use calculateDependencyVersionToInstall directly, if no suitable version directly return error, if found will continue.

Reply:
If I understand you correctly, you're suggesting that I use calculateDependencyVersionToInstall instead of validateAddonDependencies? If so, I don't believe I can do this currently because calculateDependencyVersionToInstall doesn't "validate" installed addons -- it returns the installed addon version regardless of whether the installed addon version matches the `dependency.Version`.

I will investigate some further refactoring to see if I can simplify validateAddonDependencies further.

Reply:
I refactored validateAddonDependencies to be a minimal wrapper over calculateDependencyVersionToInstall. Please review again :) .